Source/Core/VideoBackends/D3D/VertexShaderCache.cpp

```diff
@@ -195,73 +242,264 @@ void VertexShaderCache::Shutdown()
+    D3D::stateman->SetVertexShader(last_entry->shader);
+    return true;
+  }
+
+  // Background compiling?
+  if (g_ActiveConfig.CanBackgroundCompileShaders())
+  {
+    // Create a pending entry
+    VSCacheEntry entry;
+    entry.pending = true;
+    vshaders[uid] = entry;
+
+    // Queue normal shader compiling and use ubershader
+    g_async_compiler->QueueWorkItem(
+        g_async_compiler->CreateWorkItem<VertexShaderCompilerWorkItem>(uid));
+    return SetUberShader(vertex_format);
+  }
+
+  // Need to compile a new shader
+  D3DBlob* bytecode = nullptr;
+  ShaderCode code =
+      GenerateVertexShaderCode(APIType::D3D, ShaderHostConfig::GetCurrent(), uid.GetUidData());
+  D3D::CompileVertexShader(code.GetBuffer(), &bytecode);
+  if (!InsertByteCode(uid, bytecode))
+  {
+    SAFE_RELEASE(bytecode);
+    return false;
   }
 
-  ShaderCode code = GenerateVertexShaderCode(APIType::D3D, uid.GetUidData());
+  g_vs_disk_cache.Append(uid, bytecode->Data(), bytecode->Size());
+  bytecode->Release();
+  return SetShader(vertex_format);
+}
+
+bool VertexShaderCache::SetUberShader(D3DVertexFormat* vertex_format)
+{
+  D3DVertexFormat* uber_vertex_format = static_cast<D3DVertexFormat*>(
+      VertexLoaderManager::GetUberVertexFormat(vertex_format->GetVertexDeclaration()));
+  UberShader::VertexShaderUid uid = UberShader::GetVertexShaderUid();
+  if (last_uber_entry && last_uber_uid == uid)
+  {
+    if (!last_uber_entry->shader)
+      return false;
+
+    uber_vertex_format->SetInputLayout(last_uber_entry->bytecode);
+    D3D::stateman->SetVertexShader(last_uber_entry->shader);
+    return true;
+  }
 
-  D3DBlob* pbytecode = nullptr;
-  D3D::CompileVertexShader(code.GetBuffer(), &pbytecode);
+  auto iter = ubervshaders.find(uid);
+  if (iter != ubervshaders.end())
+  {
+    const VSCacheEntry& entry = iter->second;
+    last_uber_uid = uid;
+    last_uber_entry = &entry;
+
+    GFX_DEBUGGER_PAUSE_AT(NEXT_VERTEX_SHADER_CHANGE, true);
+    if (!last_uber_entry->shader)
+      return false;
+
+    uber_vertex_format->SetInputLayout(last_uber_entry->bytecode);
+    D3D::stateman->SetVertexShader(last_uber_entry->shader);
+    return true;
+  }
 
-  if (pbytecode == nullptr)
+  // Need to compile a new shader
+  D3DBlob* bytecode = nullptr;
+  ShaderCode code =
+      UberShader::GenVertexShader(APIType::D3D, ShaderHostConfig::GetCurrent(), uid.GetUidData());
+  D3D::CompileVertexShader(code.GetBuffer(), &bytecode);
+  if (!InsertByteCode(uid, bytecode))
   {
-    GFX_DEBUGGER_PAUSE_AT(NEXT_ERROR, true);
+    SAFE_RELEASE(bytecode);
     return false;
   }
-  g_vs_disk_cache.Append(uid, pbytecode->Data(), pbytecode->Size());
 
-  bool success = InsertByteCode(uid, pbytecode);
-  pbytecode->Release();
+  g_uber_vs_disk_cache.Append(uid, bytecode->Data(), bytecode->Size());
+  bytecode->Release();
+  return SetUberShader(vertex_format);
+}
+
+bool VertexShaderCache::InsertByteCode(const VertexShaderUid& uid, D3DBlob* blob)
```

Comment:
Is there a reason we need to duplicate this code?  Couldn't we just have `InsertByteCode(const ShaderUid& uid, D3DBlob* blob)`; ?